src/modules/connection.js

```diff
@@ -0,0 +1,102 @@
+var amqp = require('amqplib'),
+  assert = require('assert'),
+  packageVersion = require('../../package.json').version,
+  startedAt = new Date();
+
+var connections = {};
+
+/**
+ * Connect to the broker. We keep only 1 connection for each connection string provided in config, as advised by RabbitMQ
+ * @return {Promise} A promise that resolve with an amqp.node connection object
+ */
+function getConnection() {
+  let url = this.config.url;
+  let hostname = this.config.hostname;
+  let connection = connections[url];
+
+  //cache handling, if connection already opened, return it
+  if (connection && connection.conn) {
+    return Promise.resolve(connection.conn);
+  }
+
+  //prepare the connection internal object, and reset channel if connection has been closed
+  connection = connections[url] = { conn: null, channel: null };
+  connection.conn = amqp.connect(url, { clientProperties:
+      { hostname: hostname, bunnymq: packageVersion, startedAt: startedAt, connectedAt: new Date() }
+    }).then((conn) => {
+        //on connection close, delete connection
+        conn.on('close', () => { delete connection.conn; });
+        conn.on('error', this.config.transport.error);
+
+        connection.conn = conn;
+        return conn;
+    })
+    .catch(() => {
+      connection.conn = null;
+    });
+
+  return connection.conn;
```

Comment:
it is either a promise or a connection object

This method is used by `Promise.resolve(connect)` => to ensure it resolves with a connection, and there can be only 1 connection attempt at a time.

If I don't do that, calling `connect` multiple times will create as much connection as it has been called before the connection opening